lib/client/base.js

```diff
@@ -303,49 +314,53 @@
       info: this.options.info
     };
 
-    if (this.options.config.keyPair)
-      loginParameters.publicKey = this.options.config.keyPair.publicKey;
-
     _this.performRequest(null, 'describe', null, null, function (e, serverInfo) {
 
       if (e) return done(e);
 
-      _this.serverInfo = serverInfo;
+      var doLogin = function(){
+        _this.performRequest(null, 'login', loginParameters, null, function (e, result) {
 
-      if (_this.serverInfo.encryptPayloads && _this.clientType == 'socket') {
+          if (e) return done(e);
 
-        if (!_this.options.config.keyPair) {
-          //We generate one
-          _this.options.config.keyPair = crypto.createKeyPair();
-          loginParameters.publicKey = _this.options.config.keyPair.publicKey;
-        }
+          if (result._meta.status == 'ok') {
+            delete result._meta;
+            _this.session = result;
 
-        loginParameters = _this.__encryptLogin(loginParameters, _this.serverInfo.publicKey);
+            //write our session cookie
+            if (browser) {
+              var cookie = (result.cookieName || 'happn_token') + '=' + _this.session.token + '; path=/;';
+              if (result.cookieDomain) cookie += ' domain=' + result.cookieDomain + ';';
+              document.cookie = cookie;
+            }
+            done();
+          } else {
+            done(result.payload); // TODO: make into instanceof Error
+          }
+        });
       }
 
+      _this.serverInfo = serverInfo;
 
-      _this.performRequest(null, 'login', loginParameters, null, function (e, result) {
+      if (_this.serverInfo.encryptPayloads){
 
-        if (e) return done(e);
+        _this.__ensureCryptoLibrary(function(e){
```

Comment:
I tried this, but ran into difficulties getting it working on the browser, would rather we had functionality that worked over readability in this case. It also added more code as opposed to less code to the solution.